Build HEAD of v2.10 branch, enabling metatensor

Luthaf opened the pull request:
Creating this as a draft PR for @GiovanniBussi to comment on

This will need https://github.com/plumed/plumed2/pull/1122 and https://github.com/plumed/plumed2/pull/1123 to be merged first, and is currently pulling from a plumed fork with both PR merged. I'll update to use from plumed master later!

GiovanniBussi commented:
Thanks @Luthaf !!! Shall we have an explicit dependence on libtorch in `plumed/meta.yaml` as well?

GiovanniBussi commented:
It would also make sense to customize the label `cecam-2023` here:
`https://github.com/plumed/conda/blob/cecam-2023/.github/workflows/ci.yml`

If this contains everything, we might call it `tutorials-2024`. I would also push this on a branch with a different name rather than merging on cecam-2023 (I can do it manually)

Luthaf commented:
> Shall we have an explicit dependence on libtorch in plumed/meta.yaml as well?

There is already a transitive dependency on libtorch through metatensor-torch, but I can add an explicit one if you prefer!

> It would also make sense to customize the label cecam-2023 here

Sure, I'll customize the label!

Luthaf commented:
Is there a way to also build py-plumed from this repository? Since I would need a version with https://github.com/plumed/plumed2/pull/1129 included for the tutorial. If not, I can also try to run the simulation using some plumed-enabled LAMMPS build.

GiovanniBussi commented:
Traditionally we didn't, mostly because it was not necessary. But you can add it.

I think it makes sense if the python package for the tutorials is different from the official one. Otherwise, it should not be necessary

Luthaf commented:
So I sould copy the code from https://github.com/conda-forge/py-plumed-feedstock/ here and adapt accordingly?

Luthaf committed:
Rename label to tutorials-2024

Luthaf committed:
Build HEAD of v2.10 branch, and enable metatensor package

Luthaf committed:
Add py-plumed to the build

Luthaf committed:
Only define CONDA_BUILD_SYSROOT on CI

It prevents local builds if the macOS SDK are installed in
a different location

Luthaf commented:
Ok, this has been changed to build the HEAD of v2.10 branch, and now includes py-plumed as well.

Luthaf commented:
Updated to pull from https://github.com/plumed/plumed2, at commit https://github.com/plumed/plumed2/commit/b0ffcb55ad8cc3c71ec5c30a1bd46f68dd33bceb

GiovanniBussi commented on plumed/build.sh:
```diff
@@ -18,6 +18,10 @@ export CXXFLAGS="${CXXFLAGS//-O2/-O3}"
 
 # libraries are explicitly listed here due to --disable-libsearch
 export LIBS="-lboost_serialization -lfftw3 -lgsl -lgslcblas -llapack -lblas -lz $LIBS"
+export LIBS="-lmetatensor_torch -lmetatensor -ltorch -lc10 -ltorch_cpu $LIBS"
```
@Luthaf any specific reason for removing libraries here (boost, fftw3, gslm etc)? I suspect they won't be linked

Luthaf replied:
I did not remove anything, I added another line below. The full set of libraries will be `-lmetatensor_torch -lmetatensor -ltorch -lc10 -ltorch_cpu -lboost_serialization -lfftw3 -lgsl -lgslcblas -llapack -lblas -lz`, plus whatever was in `$LIBS` before these two lines

GiovanniBussi replied:
Sorry I didn't read correctly the script, it's fine as you did

GiovanniBussi commented on .github/workflows/ci.yml:
```diff
@@ -23,29 +23,31 @@ jobs:
-        conda-build -c conda-forge plumed
-        conda-build -c conda-forge gromacs
-        conda-build -c conda-forge lammps
+        conda-build -m .github/conda_build_config.yaml -c conda-forge plumed
```
Useful trick (`-m`), so that we have them in a single place, thanks!

Luthaf replied:
Yes! And this is intentionally not in the repository root, to prevent it from being picked on local builds: https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#conda-build-variant-config-files

GiovanniBussi commented on py-plumed/meta.yaml:
```diff
@@ -0,0 +1,52 @@
+{% set name = "py-plumed" %}
+{% set version = "2.10" %}
```
Is there a standard way to make sure the date is baked in the version in some way? What I am thinking about is if a few weeks we release 2.10 it should override this one. Or shall we just increase the build: number below?

Luthaf replied:
If you look below, the actual version ends up being `2.10.git.<hash>`. I'm not sure if this will be treated as an alpha version by conda, but I can check that. If that's the case, the future 2.10 release should have priority over the alpha release!

Luthaf replied:
Ok, I changed this to have the version be `2.10.dev1` and store the git sha in the build string instead. The whole package is then `plumed-2.10.dev1-git.b0ffcb55ad8cc3c71ec5c30a1bd46f68dd33bceb`.

This means that when 2.10 is released, it will be used instead of this one, since conda is following PEP440: https://peps.python.org/pep-0440/

Luthaf committed:
Update version/build string handling

Luthaf committed:
macos-11 builders are not provided anymore

macos-13 is the one to use for x86_64

Luthaf committed:
Do not pull binutils/hwloc on macos-arm64

Luthaf commented:
CI failed trying to pull binutils on macos (to build lammps & gromacs). Not sure why this is needed though, they don't seems to be used by the build script.

Luthaf commented:
@GiovanniBussi could you start the CI so I know if it is now running or needs more changes? Thanks!

Luthaf committed:
Do not try to deploy to anaconda when building PR

Luthaf committed:
Upgrade to the latest 2.10 commit

Luthaf committed:
Upgrade to gromacs 2023.5

Luthaf committed:
Increase Lammps build number

GiovanniBussi commented:
@Luthaf thanks! I am not 100% about the gmx update. The fact is that some method that we are using in some tutorials is not supported by gmx 2024. Is this something that you would like to include because of specific features or just because of performances? If it's the latter I would suggest to stay with an older version. Thanks!

Luthaf commented:
So the issue was that plumed v2.10 is no longer able to patch gmx 2020.4, so I updated to the latest. But if this is a problem, I can update to the oldest version supported, I seems to be `2022.5` according to `plumed-patch -l`

Luthaf commented:
Also, the build currently fails on macos-x86_64 with `configure: error: C++17 filesystem library not supported, please check your compiler`

I think this is related to the sdk version, which is fixed to 10.13 here: https://github.com/GiovanniBussi/conda-ci. Any chance you could update it? I think 10.15 should be enough, but at this point I doubt there are many users still on 10.15, and it could make sense to use 11 for both x86_64 and arm64.

GiovanniBussi commented:
I can double check if it does not break other stuff.. I guess you can also replace

./conda-ci

with

./conda-ci --sdk=11.0

Luthaf committed:
Use macos sdk version 11.0 everywhere

Luthaf committed:
Split build of different packages

Otherwise the log is too big to be viewed

Luthaf commented:
Ok, everyting builds fine on my fork, could you start CI here as well? 

I had to swap to gromacs 2023.5, since 2022.5 fails to build with very strange errors (like `FILE struct not defined`). Do the tutorials you have work with this version?

Luthaf commented:
Ah, I forgot to check the tests, and only did the build on my fork. Well, back to it!

Luthaf commented:
We ended up building the code using conda-forge infrastructure: https://github.com/metatensor/plumed-metatomic-feedstock

Thanks for your help @GiovanniBussi!
